File: backend/app/agent/investigator.py
from __future__ import annotations

import sqlite3
from collections.abc import Callable

from app.models import AgentEvent, InboxMessage, InvestigationPlan, PlanStep, ToolResult
from app.tools import bookings, compensation, guest_history, maintenance, policy_search, vision


ToolFn = Callable[[dict], ToolResult]
EventEmitter = Callable[[AgentEvent], None]


TOOL_REGISTRY: dict[str, ToolFn] = {
    "bookings.lookup": bookings.lookup,
    "maintenance.search": maintenance.search,
    "policy.search": policy_search.search,
    "guest_history.lookup": guest_history.lookup,
    "vision.verify": vision.verify,
    "compensation.evaluate": compensation.evaluate,
}


def execute_plan(
    conn: sqlite3.Connection,
    message: InboxMessage,
    plan: InvestigationPlan,
    emit: EventEmitter,
) -> list[dict]:
    evidence: list[dict] = []
    steps = list(plan.steps)
    executed = 0
    while steps and executed < 8:
        step = steps.pop(0)
        tool = TOOL_REGISTRY[step.tool]
        payload = {**step.input, "message": message.model_dump(), "evidence": evidence}
        result = tool(payload)
        evidence.append(result.model_dump())
        executed += 1
        emit(
            AgentEvent(
                event_type="tool_result",
                title=f"{step.tool}: {step.reason}",
                payload=result.model_dump(),
            )
        )
        if step.tool == "maintenance.search" and _looks_like_cluster(result) and len(steps) + executed < 8:
            steps.insert(
                0,
                PlanStep(
                    id="dynamic-ops-policy",
                    tool="policy.search",
                    reason="Maintenance evidence suggests a recurring location pattern; retrieve operations alert policy.",
                    input={"query": "repeated issue alert issue type location three tickets seven days §6.1", "top_k": 2},
                ),
            )
            emit(
                AgentEvent(
                    event_type="plan_update",
                    title="Added operations-pattern policy lookup",
                    payload={"reason": "At least three related maintenance records were found for the same area."},
                )
            )
    return evidence


def _looks_like_cluster(result: ToolResult) -> bool:
    records = result.data.get("records", [])
    issue = result.data.get("issue_type")
    by_location: dict[str, int] = {}
    for record in records:
        if record.get("issue_type") != issue:
            continue
        location = record.get("location") or record.get("room") or "property"
        by_location[location] = by_location.get(location, 0) + 1
    return any(count >= 3 for count in by_location.values())

File: pibackend/src/agent/investigator.ts
import type { AgentEvent, InboxMessage, InvestigationPlan, PlanStep, ToolResult } from "../types";
import { toolRegistry } from "../tools";

type Emit = (event: AgentEvent) => void;

export function executePlan(message: InboxMessage, plan: InvestigationPlan, emit: Emit): ToolResult[] {
  const evidence: ToolResult[] = [];
  const steps = [...plan.steps];
  let executed = 0;

  while (steps.length > 0 && executed < 8) {
    const step = steps.shift() as PlanStep;
    const tool = toolRegistry[step.tool];
    const result = tool({ ...step.input, message, evidence });
    evidence.push(result);
    executed += 1;
    emit({
      event_type: "tool_result",
      title: `${step.tool}: ${step.reason}`,
      payload: result as unknown as Record<string, unknown>
    });
    if (step.tool === "maintenance.search" && looksLikeCluster(result) && steps.length + executed < 8) {
      steps.unshift({
        id: "dynamic-ops-policy",
        tool: "policy.search",
        reason: "Maintenance evidence suggests a recurring location pattern; retrieve operations alert policy.",
        input: { query: "repeated issue alert issue type location three tickets seven days 6.1", top_k: 2 }
      });
      emit({
        event_type: "plan_update",
        title: "Added operations-pattern policy lookup",
        payload: { reason: "At least three related maintenance records were found for the same area." }
      });
    }
  }

  return evidence;
}

function looksLikeCluster(result: ToolResult): boolean {
  const records = (result.data.records as Array<Record<string, unknown>>) ?? [];
  const issueType = result.data.issue_type;
  const byLocation = new Map<string, number>();
  for (const record of records) {
    if (record.issue_type !== issueType) continue;
    const location = String(record.location ?? record.room ?? "property");
    byLocation.set(location, (byLocation.get(location) ?? 0) + 1);
  }
  return [...byLocation.values()].some((count) => count >= 3);
}
